Marlin/src/module/temperature.h

```diff
@@ -798,7 +798,7 @@ class Temperature {
         #endif
       );
       #if ENABLED(AUTO_REPORT_TEMPERATURES)
-        class AutoReportTemp : public AutoReporter<SERIAL_ALL> { void auto_report(); };
+        struct AutoReportTemp : public AutoReporter<AutoReportTemp> { void report(); };
```

Comment:
why inherit if you are using `child->report()`?

Reply:
Indeed. I think I can refactor this to just use a helper function / struct / class / lambda for the report function, no need for inclusion in the child class.

Reply:
Good question. It could be done with a static function too if you only needed one, but the code need 2 methods (temperature and SDcard).

The CRTP works this way: 
1. base class calls something it does not know until the template is specialized.
2. The child class specialize the template, so the base class is base<child>, now the base class can static_cast to Child safely, it's known so `static_cast<Child*>(this)->` will call the child class's method
3. There is no need the child method is called the same as the base class, it's not a virtual table. If it's called the same, it'll work as well (since you explicitly call the child's method) but it's confusing to the reader that'll think it's a virtual table (IMHO). So unless it's absolutely required, I'm not naming the child method the same.

From the compiler point of view, auto_report from base is just a method that's calling another method of itself. There is no runtime dispatch, it's a compile-time dispatch. The main advantage is that you've a single interface in the base and the "child" is only implementing the feature you need (just like a pure virtual class without the cost).

Reply:
Yes, we can have 
```
struct AutoReportTemp { void report(); };
AutoReporter<AutoReportTemp> auto_reporter;


void Temperature::AutoReportTemp::report() {
       print_heater_states(active_extruder);
       SERIAL_EOL();
     }
```

And don't use inheritance at all... but I thought you wanted to keep the pattern, so I "got it"....

Reply:
CRTP is only need when we have a base and derived depending on each other interface. But in the current code, only Base need to call one method from child... So the child don't need share any special interface with parent, at all... I wonder... Just thinking loud...

Reply:
Yep that would work if the report method was static and does not need anything from AutoReporter (like in both usage here, you're 100% right).

Reply:
I've not checked what's in actual report method, sorry.

Reply:
Indeed, the child class does not need anything from the base class, I'll remove the inheritance.

Reply:
CRTP is cute. We can leave it. Now @thinkyhead should decide what he like most: a cute CRTP or avoid inheritance when not needed... 🤓

Reply:
the report method can be static too.